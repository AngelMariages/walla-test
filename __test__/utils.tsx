import { render } from '@testing-library/react';
import {
    FavoritesContext,
    FavoritesContextType,
} from 'context/FavoritesContext';
import { FiltersContext, FiltersContextType } from 'context/FiltersContext';

type ContextValues = {
    filtersContext?: Partial<FiltersContextType>;
    favoritesContext?: Partial<FavoritesContextType>;
};

export const renderWithContext = (
    ui: React.ReactElement,
    contextValues?: ContextValues
) => {
    return render(
        <FiltersContext.Provider
            value={{
                search: '',
                setSearch: jest.fn(),
                page: 1,
                incrementPage: jest.fn(),
                sort: {
                    sortBy: 'title',
                    sortOrder: 'asc',
                },
                setSort: jest.fn(),
                ...(contextValues?.filtersContext || {}),
            }}>
            <FavoritesContext.Provider
                value={{
                    favorites: [],
                    favoritesVisible: false,
                    addFavorite: jest.fn(),
                    removeFavorite: jest.fn(),
                    isFavorite: jest.fn(),
                    toggleFavorites: jest.fn(),
                    ...(contextValues?.favoritesContext || {}),
                }}>
                {ui}
            </FavoritesContext.Provider>
        </FiltersContext.Provider>
    );
};
