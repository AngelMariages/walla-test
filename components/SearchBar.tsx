'use client';

import { useState } from 'react';
import SearchIcon from 'public/icons/search.svg';
import { usePathname, useRouter } from 'next/navigation';

const getSearchFromPathname = (pathname: string | null) => {
    if (!pathname) {
        return '';
    }

    const match = pathname.match(/\/search\/(.*)/);
    return match ? match[1] : '';
};

const Search: React.FC = () => {
    const router = useRouter();
    const pathName = usePathname();
    const [search, setSearch] = useState(getSearchFromPathname(pathName));

    const handleSearch = async (e: React.FormEvent<HTMLFormElement>) => {
        e.preventDefault();

        if (!search) {
            router.push('/', {
                forceOptimisticNavigation: true,
            });
            return;
        }

        router.push(`/search/${search}`, {
            forceOptimisticNavigation: true,
        });
    };

    return (
        <form onSubmit={handleSearch}>
            <div className="ml-6 md:ml-8 flex gap-4 md:gap-2">
                <input
                    className="rounded-md border border-gray-300 px-2 py-1"
                    type="text"
                    placeholder="Search"
                    value={search}
                    onChange={(e) => setSearch(e.target.value)}
                />
                <button type="submit">
                    <SearchIcon alt="search" />
                </button>
            </div>
        </form>
    );
};

export default Search;
