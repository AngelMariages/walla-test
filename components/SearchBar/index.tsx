import { debounce } from 'lib/debounce';
import { forwardRef, useMemo } from 'react';
import SearchIcon from 'public/icons/search.svg';

type Props = {
    setSearch: (search: string) => void;
    onSubmit?: (e: React.FormEvent<HTMLFormElement>) => void;
};

export default forwardRef<HTMLInputElement, Props>(function SearchBar(
    { setSearch, onSubmit },
    searchRef
) {
    const debouncedSetSearch = useMemo(
        () => debounce(setSearch, 500),
        [setSearch]
    );

    return (
        <form onSubmit={onSubmit} data-testid="search-form">
            <div className="ml-4 md:ml-8 flex gap-2 md:gap-4">
                <input
                    data-testid="search-input"
                    ref={searchRef}
                    className="rounded-md border border-gray-300 px-2 py-1"
                    type="text"
                    placeholder="Search"
                    onChange={(e) => {
                        debouncedSetSearch(e.target.value);
                    }}
                />
                <button type="submit" className="ml-2">
                    <SearchIcon alt="search" />
                </button>
            </div>
        </form>
    );
});
