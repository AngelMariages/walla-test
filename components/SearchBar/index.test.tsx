import { fireEvent, act } from '@testing-library/react';
import { renderWithContext } from '__test__/utils';
import SearchBar from '.';

const mockRouterPush = jest.fn();

jest.mock('next/router', () => ({
    useRouter: () => ({
        push: mockRouterPush,
    }),
}));

jest.mock('next/navigation', () => ({
    usePathname: () => '/search/test',
}));

describe('SearchBar', () => {
    it('should render correctly', () => {
        const { container } = renderWithContext(<SearchBar />);
        expect(container).toMatchSnapshot();
    });

    it('should call onSearch when input changes', async () => {
        const onSearch = jest.fn();
        const { getByTestId } = renderWithContext(<SearchBar />, {
            filtersContext: {
                setSearch: onSearch,
            },
        });
        const input = getByTestId('search-input');

        await act(() => {
            fireEvent.change(input, { target: { value: 'test' } });
        });
        expect(onSearch).toHaveBeenCalledTimes(1);
        expect(onSearch).toHaveBeenCalledWith('test');
    });

    it('should call router.push when form is submitted', async () => {
        const { getByTestId } = renderWithContext(<SearchBar />);
        const input = getByTestId('search-input');
        const form = getByTestId('search-form');

        await act(() => {
            fireEvent.change(input, { target: { value: 'second-test' } });

            fireEvent.submit(form);
        });
        expect(mockRouterPush).toHaveBeenCalledWith(
            '/search/second-test',
            undefined,
            {
                shallow: true,
            }
        );
    });
});
