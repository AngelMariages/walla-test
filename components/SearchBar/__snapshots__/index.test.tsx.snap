// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBar should render correctly 1`] = `
<div>
  <form
    data-testid="search-form"
  >
    <div
      class="ml-4 md:ml-8 flex gap-2 md:gap-4"
    >
      <input
        class="rounded-md border border-gray-300 px-2 py-1"
        data-testid="search-input"
        placeholder="Search"
        type="text"
      />
      <button
        class="ml-2"
        type="submit"
      >
        <svg-url
          alt="search"
        />
      </button>
    </div>
  </form>
</div>
`;
